src/backend/distributed/relay/relay_event_utility.c

```diff
@@ -120,14 +121,23 @@ RelayEventExtendNames(Node *parseTree, char *schemaName, uint64 shardId)
 				{
 					Constraint *constraint = (Constraint *) command->def;
 					char **constraintName = &(constraint->conname);
+					const bool rvMissingOk = false;
+					relationId = RangeVarGetRelid(alterTableStmt->relation,
+												  AccessShareLock,
+												  rvMissingOk);
 
 					if (constraint->contype == CONSTR_PRIMARY && constraint->indexname)
 					{
 						char **indexName = &(constraint->indexname);
 						AppendShardIdToName(indexName, shardId);
 					}
 
-					AppendShardIdToName(constraintName, shardId);
+					/* append shardId to constraint names when necessary */
+					if (!PartitionedTable(relationId) ||
```

Comment:
Two questions at first glance:

- Would this work for the existing partitioned tables on the cluster?
- Would this work with drop/alter constraints?

Reply:
> Would this work for the existing partitioned tables on the cluster?

Yes, and no.

My changes here only affect `ALTER TABLE .. ADD CONSTRAINT` queries. We no longer will append shardId to constraint names after an upgrade.

If on an earlier version, the user added a new constraint to a partitioned table after distributing it, the constraint names on the shards will contain shardIds and this can block new partition creation. However if the user drops said constraints, upgrades, and recreates those constraints, the constraint names will be different on worker nodes, and it will work for them.

Just to be safe, I will do some testing on this scenario.

> Would this work with drop/alter constraints?

I guess so. Let me add some tests to make sure.

Reply:
Only `ADD|DROP|VALIDATE CONSTRAINT` queries are supported, and we error out on `ALTER TABLE .. ALTER CONSTRAINT`.

`DROP CONSTRAINT` works as expected. I will add a regression test for this.

Reply:
added a test case in b8736fb

Reply:
I think we should consider the first part a little bit more. Because say that we have this patch, then do the followings:

```sql
CREATE TABLE part_table (
    work_ymdt timestamp without time zone NOT NULL,
    seq bigint NOT NULL,
    my_seq bigint NOT NULL,
    work_memo character varying(150),
    CONSTRAINT work_memo_check CHECK ((octet_length((work_memo)::text) <= 150))
)
PARTITION BY RANGE (work_ymdt);

--2. perform create_distributed_table
SELECT create_distributed_table('part_table', 'seq');

--3. add a partition
CREATE TABLE part_table_p202008 PARTITION OF part_table  FOR VALUES FROM ('2020-08-01 00:00:00') TO ('2020-09-01 00:00:00');

--4. add a check constraint
alter table part_table add CONSTRAINT my_seq CHECK (my_seq > 0);
```

Then add this patch, and try:

```sql
[local] talha@talha:9700-18814=# CREATE TABLE part_table_p202009 PARTITION OF part_table  FOR VALUES FROM
 ('2020-09-01 00:00:00') TO ('2020-10-01 00:00:00');
ERROR:  child table is missing constraint "my_seq_102008"
CONTEXT:  while executing command on localhost:9701
Time: 39.846 ms
[local] talha@talha:9700-18814=#
```

so certainly we wouldn't want users to see this error, it seems that adding a new constraint and dropping the existing one seems to be working but we should possibly cover the above case. Once we find a solution, we can possibly add an upgrade test to see that it indeed works. 

I would do the following if possible:

- Add an upgrade test to see that the above indeed gets an error with the current solution
- In local, play a bit with different possibilities to see if something else also errors so that the solution covers that as well (alter/drop was working fine in my case)
- Try to find a solution that fixed the failure in the test

If adding an upgrade test isn't trivial you can do that at the end as well.

Reply:
We discussed the hardships with upgrade test scenarios with Talha offline.

- In mixed mode, the workers in earlier versions will still hit the issue. If we create alternative test outputs, then we would not know if only the workers with older citus versions hit the issue or all of them. That's why it will not be easy to catch regressions.
- In Mac, the upgrade test mechanism is broken due to several issues. It is not possible to create and test an upgrade test until it is fixed. This is clearly out of scope for this PR
  - various permission issues including unallowed /tmp/ directory accesses.

I added a migration script and tested it manually.

Reply:
> In Mac, the upgrade test mechanism is broken due to several issues. It is not possible to create and test an upgrade test until it is fixed. This is clearly out of scope for this PR

I think it  is very important to have  this test in the regression files. Can  we use an Azure VM or such to add the test?


> In mixed mode, the workers in earlier versions will still hit the issue. If we create alternative test outputs, then we would not know if only the workers with older citus versions hit the issue or all of them. That's why it will not be easy to catch regressions.

I think it is OK to not have regression tests for this change

Reply:
>I think it is OK to not have regression tests for this change

The current upgrade test mechanism uses the same schedule in both mixed mode and normal mode. See https://github.com/citusdata/citus/blob/61bf2fb4774790bcd4ef929b03bb2a8965ff511f/.circleci/config.yml#L398-L408

To make sure we use a different schedule on different modes, we can make some changes to the upgrade test mechanism, in a separate PR maybe.

Reply:
Answering the original questions again:

> Would this work for the existing partitioned tables on the cluster?

Yes it works. See the upgrade tests.

> Would this work with drop/alter constraints?

I added tests for ADD and DROP and they work. VALIDATE also works, though I forgot to add a regression test for it. The other `ALTER CONSTRAINT` queries are not supported by Citus.

```sql
postgres# ALTER TABLE part_table ALTER CONSTRAINT my_seq DEFERRABLE;
ERROR:  0A000: alter table command is currently unsupported
DETAIL:  Only ADD|DROP COLUMN, SET|DROP NOT NULL, SET|DROP DEFAULT, ADD|DROP|VALIDATE CONSTRAINT, SET (), RESET (), ATTACH|DETACH PARTITION and TYPE subcommands are supported.
```

Is this ok to resolve this thread @onderkalaci ?

Reply:
Yes, I think so. Thanks for the upgrade tests!